### ISSUE TITLE
[Python] Support for the free-threaded build of CPython 3.13

### ISSUE THREAD

**Issue by lysnikolaou** [apache/arrow#43536]
### Describe the enhancement requested

Hey all! 👋

I've started looking into adding support for the 3.13 free-threaded build of CPython. This aims to be a tracking issue for all work necessary to support it. A very high-level list of steps is (more details to follow as I investigate more issues):

- [x] Set up CI for free-threading (This needs to be done for 3.13 as well as part of #43519).
  - [x] https://github.com/apache/arrow/pull/43671 
- [x] Move away from APIs that use borrowed references (`PyList_GetItem`, `PyDict_GetItem` and friends) under the free-threaded build
  - [x] https://github.com/apache/arrow/pull/43540
- [x] Audit C extension modules for thread-safety issues
  - [x] https://github.com/apache/arrow/issues/44046
- [ ] `pyarrow` is a good candidate to test Cython again, since it's got lots of Cython code. Ideally test Cython nightly and also cython/cython#6226.
- [x] Mark C/Cython extension modules as thread-safe with the `Py_mod_gil` slot (or `PyUnstable_Module_SetGIL` / `freethreading_compatible`)
  - [x] https://github.com/apache/arrow/pull/43606
- [x] Upload (nightly) wheels for the free-threaded build.
  - [x] https://github.com/apache/arrow/issues/43964

Hope this all sounds good! Let me know in case there's something that doesn't seem right.

### Component(s)

Python

**Comment by jorisvandenbossche** [comment]
@lysnikolaou for adding a free-threaded test build, I suppose we will want to add a wheel build as well (so after https://github.com/apache/arrow/pull/43539 is finally ready), to ensure we have nightly wheels people can test with? 
But regardless of wheels, it might be good to have a plain test build as well, which could mimic the cpython-debug build that Antoine added? (https://github.com/apache/arrow/pull/43565) That is something that doesn't have to wait on the wheels / general python 3.13 support in CI, I think

**Comment by lysnikolaou** [comment]
> But regardless of wheels, it might be good to have a plain test build as well, which could mimic the cpython-debug build that Antoine added? (#43565) That is something that doesn't have to wait on the wheels / general python 3.13 support in CI, I think

On it.

**Comment by kou** [comment]
Issue resolved by pull request 43606
https://github.com/apache/arrow/pull/43606

**Comment by kou** [comment]
@lysnikolaou Could you create a sub issue per sub item? Our release management scripts don't support multiple PRs per issue for now.

**Comment by jorisvandenbossche** [comment]
This is fine to keep as the general tracking issue (the release management script will just only list this issue in the changelog, which IMO is perfectly fine)

**Comment by lysnikolaou** [comment]
I agree that this issue should be kept open, since it provides a nice overview of what level of support can be expected.

@kou Sure! Should I open subissues for the already open/merged PRs as well or is it okay to just do it in the future?

**Comment by kou** [comment]
If @jorisvandenbossche doesn't like create subissues, I don't ask you to create subissues strongly. (As a committer, I don't want to reopen this issue manually after I merge a PR by our merge script.)
@raulcd What do you think about this as a release manager? We can't assign multiple milestones to one issue. If we slip some subtasks to 19.0.0, it may cause a problem.

FYI: We used an umbrella issue and subissues when we added support for Azure filesystem: https://github.com/apache/arrow/issues/18014

**Comment by raulcd** [comment]
At the moment all our merge (closes issues automatically) and release scripts (generation of release notes, cherry picking) expect one issue per PR.

This was a requirement for JIRA but now that we are tracking issues on GH I think we can revisit that pattern, we could even avoid some issues and just use PRs for some things but this would require some time to adapt our current scripts.

As per this particular issue, the biggest potential problems for the release:
- As you have already seen the issue will keep getting closed every time we merge a subtask.
- PRs merged pointing to this PR will probably not appear on the release notes (as far as I know release notes are generated from issues tagged with the correct milestone not PRs)
- If we merge something after the feature freeze it won't get cherry-picked by our scripts

To be honest not really breaking issues for the release but if this is a workflow we want to use (which I think we should) we should adapt our scripts. I am currently on vacation so might take some time responding.

**Comment by raulcd** [comment]
> @kou Sure! Should I open subissues for the already open/merged PRs as well or is it okay to just do it in the future?

Already merged PRs point to the "parent" issue on the commit message, this is what is used to link the commits to issues on the release scripts, unless we want to change the commit history (which we are not going to do) it doesn't matter if we open issues for the ones already merged :)

BTW, thanks for the great work you are doing @lysnikolaou

**Comment by lysnikolaou** [comment]
Ok then, I'll be creating subissues for PRs I open from now on. Thanks everyone!

**Comment by pitrou** [comment]
Issue resolved by pull request 43540
https://github.com/apache/arrow/pull/43540

**Comment by lysnikolaou** [comment]
Can we please reopen this? These are PRs that I had created before we discussed this. Sorry!

**Comment by pitrou** [comment]
Woops, sorry, I've been a bit too optimistic here :)

**Comment by pitrou** [comment]
Thanks to https://github.com/indygreg/python-build-standalone/pull/326 , there should be very soon free-threaded builds available on https://github.com/indygreg/python-build-standalone/ . That might be easier than using a PPA and being constrained to Ubuntu.

**Comment by raulcd** [comment]
@jorisvandenbossche @pitrou is there anything to be done here for 18.0.0?

**Comment by raulcd** [comment]
I am closing this as done for 18.0.0. The only sub-task missing is around testing Cython. Please feel free to open a new issue for that if necessary.
